clusterloader2/pkg/measurement/common/api_availability_measurement.go

```diff
@@ -0,0 +1,185 @@
+	result.StatusCode(&status)
+	a.updateAvailabilityMetrics(status == 200, a.clusterLevelMetrics)
+}
+
+func (a *apiAvailabilityMeasurement) start(config *measurement.Config, SSHToMasterSupported bool, probeDuration int) error {
+	a.hosts = config.ClusterFramework.GetClusterConfig().MasterIPs
+	if len(a.hosts) < 1 {
+		return goerrors.Errorf("APIAvailability measurement can't start due to lack of master IPs")
+	}
+
+	k8sClient := config.ClusterFramework.GetClientSets().GetClient()
+	provider := config.ClusterFramework.GetClusterConfig().Provider
+
+	a.isRunning = true
+	a.stopCh = make(chan struct{})
+	a.wg.Add(1)
+
+	go func() {
+		defer a.wg.Done()
+		for {
+			select {
+			case <-a.stopCh:
+				return
+			case <-time.After(time.Duration(probeDuration)):
+				a.updateClusterAvailabilityMetrics(k8sClient)
+				if SSHToMasterSupported {
+					a.updateMasterAvailabilityMetrics(k8sClient, config, provider)
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+// Execute starts the api-server healthz probe end point from start action and
+// collects availability metrics in gather.
+func (a *apiAvailabilityMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	SSHToMasterSupported := !config.ClusterFramework.GetClusterConfig().Provider.Features().SupportSSHToMaster
+
+	if !SSHToMasterSupported {
+		klog.Infof("ssh to master nodes not supported. Measurement would have only Cluster Level Metrics")
+	}
+
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	probeFrequency, err := util.GetIntOrDefault(config.Params, "frequency", 1)
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if a.isRunning {
+			klog.Infof("%s: measurement already running", a)
+			return nil, nil
+		}
+		return nil, a.start(config, SSHToMasterSupported, probeFrequency)
+	case "gather":
+		err := a.stopAndSummarize(SSHToMasterSupported, probeFrequency)
+		if err != nil {
+			return nil, err
+		}
+		return a.summaries, nil
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (a *apiAvailabilityMeasurement) createClusterAvailabilitySummary(probeFrequency int) {
+	clusterAvailabilityMetricsSummary := a.clusterLevelMetrics.buildAPIAvailabilityMetricsSummary(probeFrequency, true, "")
+	a.apiAvailabilityOutput.ClusterMetrics = clusterAvailabilityMetricsSummary
+}
+
+func (a *apiAvailabilityMeasurement) createMasterAvailabilitySummary(probeFrequency int) {
+	for _, host := range a.hosts {
+		masterHostAvailabilitySummary := a.hostLevelMetrics[host].buildAPIAvailabilityMetricsSummary(probeFrequency, false, host)
+		a.apiAvailabilityOutput.MasterMetrics = append(a.apiAvailabilityOutput.MasterMetrics, masterHostAvailabilitySummary)
+	}
+}
+
+func (a *apiAvailabilityMeasurement) stopAndSummarize(SSHToMasterSupported bool, probeFrequency int) error {
+	if !a.isRunning {
+		return nil
+	}
+	close(a.stopCh)
+	a.wg.Wait()
+
+	a.createClusterAvailabilitySummary(probeFrequency)
```

Comment:
Given apiAvailabilityOutput shouldn't be a measurement-level field, let's simply do:
```
output := apiAvailabilityOutput{}
output.createClusterSummary(a.clusterLevelMetrics, probeFrequency)
if SSHToMasterSupported {
  output.createMastersSummary(a.hostLevelMetrics, probeFrequency)
}
```

[which also means removing the createSummary methods above and making the output methods]